### PR TITLE
allow caller to pass in an already encoded attachment

### PR DESCRIPTION
**Pull request by john-cai**

### CODE REVIEW

File: gomail_test.go [hunk 1]
```diff
@@ -302,6 +302,44 @@ func TestAttachmentsOnly(t *testing.T) {
...
+	msg.SetHeader("From", "from@example.com")
+	msg.SetHeader("To", "to@example.com")
+	content1 := "hello"
+	content2 := "goodbye"
+	content1Buf := make([]byte, base64.StdEncoding.EncodedLen(len(content1)))
+	content2Buf := make([]byte, base64.StdEncoding.EncodedLen(len(content2)))
+	base64.StdEncoding.Encode(content1Buf, []byte(content1))
+	base64.StdEncoding.Encode(content2Buf, []byte(content2))
+	msg.Attach(CreateEncodedFile("test.pdf", content1Buf))
+	msg.Attach(CreateEncodedFile("test.zip", content2Buf))
```
Comment (kanekv) [inline gomail_test.go]: attach unencoded file for completeness

File: export.go [hunk 2]
```diff
@@ -152,7 +152,7 @@ func (w *messageWriter) addFiles(files []*File, isAttachment bool) {
 			}
 		}
 
-		w.write(h, f.Content, Base64)
+		w.write(h, f.Content, f.Encoding)
```
Comment (dtjm) [inline export.go]: Do you need to check that this is either Base64 or Base64Encoded?

File: gomail.go [hunk 3]
```diff
@@ -100,6 +100,8 @@ const (
 	// Unencoded can be used to avoid encoding the body of an email. The headers
 	// will still be encoded using quoted-printable encoding.
 	Unencoded Encoding = "8bit"
+	// Base64Encoded represents data that has already been base64 encoded
+	Base64Encoded Encoding = "base64encoded"
```
Comment (dtjm) [inline gomail.go]: Will this string ever be used as output in the mail message? 

Also the naming is a bit confusing.  E.g. Base64 Encoding and Base64Encoded Encoding sound the same.  Maybe call it Base64PreEncoded and Base64PostEncoded or something to differentiate.  Also please update the comment on Base64 to indicate that the base64 encoding will be performed by gomail if that Encoding is used.

File: gomail_test.go [hunk 4]
```diff
@@ -302,6 +302,55 @@ func TestAttachmentsOnly(t *testing.T) {
...
+	msg.Attach(file1)
+	msg.Attach(file2)
+	msg.Attach(file3)
+
+	want := message{
+		from: "from@example.com",
+		to:   []string{"to@example.com"},
+		content: "From: from@example.com\r\n" +
+			"To: to@example.com\r\n" +
+			"Content-Type: multipart/mixed; boundary=_BOUNDARY_1_\r\n" +
```
Comment (dtjm) [inline gomail_test.go:339]: ~~Is there a need for a test that produces multipart/alternative, or somehow includes text/plain and text/html content?~~

Looks like existing tests cover this already.

**Comment by dtjm** [comment]
Can you add a comment here as to why this is Base64 and not f.encoding, or else someone might try to change it in the future.

https://github.com/Kane-Sendgrid/gomail/pull/4/files#diff-a553d2c5d29ef25a8e160b082621dbbfR143

**Comment by dtjm** [comment]
CR 1 √. Do lots of testing and share your test plan / results please.

**Comment by Dharmik-Patel** [comment]
@john-cai why is this not merged? did you get back to @dtjm on test plan and results? is there anything needed for this to get merged? Looks like you have merged the attachment story to master without this merging to master. This is causing godep restore to not work

**Pull request merged**